### PR TITLE
Don't change global FPU state during round-half-to-even

### PR DESCRIPTION
**Pull request by tiagoshibata**
**Description**: Do round-half-to-even without affecting global FPU state. There's no need to change global FPU state to do the operation. The change in global FPU state was also being flagged by the osmode CRT.

**Motivation and Context**
- Stop modifying global program state
- Make it build with the osmode CRT

### CODE REVIEW

File: onnxruntime/core/util/qmath.h [hunk 1]
```diff
@@ -51,9 +51,7 @@ void QGemm(
     concurrency::ThreadPool* thread_pool);
 
 inline float RoundHalfToEven(float input) {
-  std::fesetround(FE_TONEAREST);
-  auto result = std::nearbyintf(input);
-  return result;
+  return input - std::remainderf(input, 1.f);
```
Comment (tiagoshibata) [inline onnxruntime/core/util/qmath.h:54]: From cppreference:

`The IEEE floating-point remainder of the division operation x/y calculated by this function is exactly the value x - n*y, where the value n is the integral value nearest the exact value x/y. When |n-x/y| = ½, the value n is chosen to be even.`

**Review by ryanlai2 — APPROVED** [review]

**Review by tracysh — DISMISSED** [review]

**Review by yufenglee — APPROVED** [review]

**Comment by tiagoshibata** [comment]
Added a small change to handle infinity properly (previously remainder would return NaN and the entire result would be NaN)

File: onnxruntime/core/util/qmath.h (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/core/util/qmath.h:58]: What's the performance impact? Is it slower?

Comment (tiagoshibata) [reply onnxruntime/core/util/qmath.h:58]: Following program:

```cpp
#include <cfenv>
#include <cmath>

inline float RoundHalfToEven(float input) {
    if (!std::isfinite(input)) {
        return input;
    }
    return input - std::remainderf(input, 1.f);
}

inline float RoundHalfToEven2(float input) {
    std::fesetround(FE_TONEAREST);
    auto result = std::nearbyintf(input);
    return result;
}

int main() {
    float result{};
    for (float x = 1; x < 3e+038; x *= 1.0000001) {
        result += RoundHalfToEven2(x);  // or RoundHalfToEven
    }
    return static_cast<int>(result);
}
```

Gives me:

|       | Previous | Proposed |
|-------|----------|----------|
| MSVC  | 41.928   | 18.527   |
| clang | 40.267   | 17.455   |

So new method is at least 2.3x faster on both compilers on my CPU.

**Review by ryanlai2 — APPROVED** [review]

**Review by tracysh — APPROVED** [review]

**Review by yufenglee — APPROVED** [review]

**Comment by tiagoshibata** [comment]
Do you know what the segfault in the training pipeline is about?

**Pull request merged**